data/KMeansClustering.py

```diff
@@ -0,0 +1,221 @@
+		List of ``n_features``-dimensional data points. Each row corresponds
+		to a single data point.
+	labels : array-like, shape (``n_samples``,)
+		Predicted labels for each sample.
+	Returns
+	-------
+	score: float
+		The resulting Davies-Bouldin score.
+	References
+	----------
+	.. [1] Davies, David L.; Bouldin, Donald W. (1979).
+	   `"A Cluster Separation Measure"
+	   <https://ieeexplore.ieee.org/document/4766909>`__.
+	   IEEE Transactions on Pattern Analysis and Machine Intelligence.
+	   PAMI-1 (2): 224-227
+	"""
+	X, labels = check_X_y(X, labels)
+	le = LabelEncoder()
+	labels = le.fit_transform(labels)
+	n_samples, _ = X.shape
+	n_labels = len(le.classes_)
+	check_number_of_labels(n_labels, n_samples)
+
+	intra_dists = np.zeros(n_labels)
+	centroids = np.zeros((n_labels, len(X[0])), dtype=np.float)
+	for k in range(n_labels):
+		cluster_k = safe_indexing(X, labels == k)
+		centroid = cluster_k.mean(axis=0)
+		centroids[k] = centroid
+		intra_dists[k] = np.average(pairwise_distances(
+			cluster_k, [centroid]))
+
+	# centroid_distances will contain zeros in the diagonal
+	centroid_distances = pairwise_distances(centroids)
+
+	if np.allclose(intra_dists, 0) or np.allclose(centroid_distances, 0):
+		return 0.0
+
+	# Compute score avoiding division by zero by adding an epsilon
+	# this leads to high values in the diagonal's result
+	score = (intra_dists[:, None] + intra_dists) / (centroid_distances + 1e-15)
+
+	# Simply put the diagonal to zero
+	score[np.eye(centroid_distances.shape[0]) == 1] = 0
+
+	# Here is the original code
+	# score = (intra_dists[:, None] + intra_dists) / (centroid_distances)
+	# score[score == np.inf] = np.nan
+	return np.mean(np.nanmax(score, axis=1))
+
+
+class KMeansClustering(CustomData):
+	@staticmethod
+	def create_data(X: dt.Frame = None) -> Union[str, List[str],
+												 dt.Frame, List[dt.Frame],
+												 np.ndarray, List[np.ndarray],
+												 pd.DataFrame, List[pd.DataFrame]]:
+		if X is None:
+			return []
+
+		# check the datatype of user-defined columns
+		if not isinstance(include_columns, list):
+			raise ValueError("Variable: 'include_columns' should be <list>")
+		if not isinstance(ignore_columns, list):
+			raise ValueError("Column: 'ignore_columns' should be <list>")
+		if not isinstance(num_clusters, int):
+			raise ValueError("Column: 'num_clusters' should be <int>")
+
+		## validate user-inputs and override the columns given by user
+		features = list(X.names)	
+		if len(include_columns) > 0:
+			for _ in include_columns:
+				if _ not in list(X.names):
+					raise ValueError("Column: '"+str(_) + "' is not present in the dataset")
+			features = include_columns
+		
+		## list to ignore specific columns given by user
+		features = [_f for _f in features if _f not in ignore_columns]
+
+		## handle columns with missing values 
+		ignore_ = [] 
+		X_df = X.to_pandas()
+		for col in features:
+			if X_df[col].dtype == "object": # ignore categorical columns
```

Comment:
What I would do here is go simple label encoding for string categoricals.
```
    X_df[f"{col}_enc"] = LabelEncoder().fit_transform(X_df[col].to_numpy())
    del X_df[:, col]
```
No tested but should work, not ideal but at least rows sharing the same values are recognized by Kmeans.

Reply:
I tested that same logic for categorical variables this morning. It was working well. But later removed it as many people online have suggested that there might be problems with the label encoded representation. Different representation may result in different clusters. 

Additionally, a variant of K-means ie. KModes might be a good choice. 

https://datascience.stackexchange.com/questions/22/k-means-clustering-for-mixed-numeric-and-categorical-data

Reply:
Sure this is not an ideal representation but it's something.

I think we could work on a PR in closed recipes to find better ways for categoricals and have LabelEncoding for now.

Reply:
I never tried sparse representation for Kmeans and maybe H2O-3 has better support for that as well.

Reply:
Agreed, Added the label encoding for categorical variables.